aten/src/ATen/native/quantized/cpu/q_avgpool.cpp

```diff
@@ -0,0 +1,202 @@
+
+          /* local pointers */
+          scalar_t *ip = input_p   + d*istrideD + istartH*istrideH + istartW*istrideW;
+          scalar_t *op = output_p  + d*osizeH*osizeW + oh*osizeW + ow;
+
+          /* compute local average: */
+          int64_t sum = 0;
+          int ih, iw;
+          for(ih = 0; ih < kH; ih++) {
+            for(iw = 0; iw < kW; iw++) {
+              int64_t val = (ip + ih*istrideH + iw*istrideW)->val_;
+              sum += val;
+            }
+          }
+
+          /* set output to local average */
+          op->val_ = static_cast<underlying_t>(std::nearbyint(sum * kHWr));
+        }
+      }
+    }
+  });
+}
+
+template <typename scalar_t, typename underlying_t>
+void adaptive_avg_pool2d_out_frame(scalar_t *input_p,
+                                   scalar_t *output_p,
+                                   int64_t sizeB,
+                                   int64_t sizeD,
+                                   int64_t isizeH,
+                                   int64_t isizeW,
+                                   int64_t osizeH,
+                                   int64_t osizeW,
+                                   int64_t istrideB,
+                                   int64_t istrideD,
+                                   int64_t istrideH,
+                                   int64_t istrideW) {
+  at::parallel_for(0, sizeB, 0, [&](int64_t start, int64_t end) {
+    for (auto b = start; b < end; b++) {
+      adaptive_avg_pool2d_single_out_frame<scalar_t, underlying_t>(
+        input_p + b * istrideB,
+        output_p + b * sizeD * osizeH * osizeW,
+        sizeD,
+        isizeH, isizeW,
+        osizeH, osizeW,
+        istrideD,
+        istrideH, istrideW);
+    }
+  });
+}
+
+void adaptive_avg_pool2d_out_template(Tensor& output, Tensor input,
+                                      IntArrayRef output_size) {
+  for (int64_t i = 0; i < input.dim(); i++) {
+    TORCH_CHECK(input.size(i) > 0,
+      "adaptive_avg_pooling2d(): expected input to have non-empty spatial "
+      "dimensions, but input has sizes ", input.sizes(), " with dimension ", i,
+      " being empty");
+  }
+
+  TORCH_CHECK((input.dim() == 3 || input.dim() == 4),
+    "non-empty 3D or 4D (batch mode) tensor expected for input");
+
+  /* sizes */
+  int64_t sizeD  = input.size(-3);
+  int64_t isizeH = input.size(-2);
+  int64_t isizeW = input.size(-1);
+  /* strides */
+  int64_t istrideD = input.stride(-3);
+  int64_t istrideH = input.stride(-2);
+  int64_t istrideW = input.stride(-1);
+
+  auto osizeH = output_size[0];
+  auto osizeW = output_size[1];
+
+  /* resize output */
+  std::vector<int64_t> output_sizes;
+  int64_t sizeB = 0;
+  if (input.dim() == 3) {
+    output_sizes = {sizeD, osizeH, osizeW};
+  } else {
+    sizeB = input.size(-4);
+    output_sizes = {sizeB, sizeD, osizeH, osizeW};
+  }
+  if (!output.is_quantized() || output.sizes() != output_sizes) {
+    output = at::_empty_affine_quantized(output_sizes, input.options(),
```

Comment:
hm, is it semantics we want? if the output the wrong size - inherit quantizer from input, otherwise from output? it's not how majority of ops work in PyTorch and I don't think it's a good behavior.

probably a better approach is to force output to be already quantized and error out if the size is wrong. @gchanan

Reply:
So basically throw an error at this point?

Reply:
yeah, if output is provided, then it must be quantized and has the correct size.